File: classifier/data/raw/human/samples/project_euler_zig/problem_044.zig
const std = @import("std");

// Project Euler 44: smallest difference of pentagonal numbers.
pub fn main() !void {
    var best: u64 = std.math.maxInt(u64);
    var j: u64 = 2;
    while (j <= 5_000) : (j += 1) {
        const pj = pentagonal(j);
        var k: u64 = 1;
        while (k < j) : (k += 1) {
            const pk = pentagonal(k);
            const difference = pj - pk;
            if (difference < best and isPentagonal(difference) and isPentagonal(pj + pk)) {
                best = difference;
            }
        }
    }
    printNumber(best);
}

fn pentagonal(n: u64) u64 {
    return n * (3 * n - 1) / 2;
}

fn isPentagonal(n: u64) bool {
    const discriminant = 1 + 24 * n;
    var root: u64 = @intFromFloat(std.math.sqrt(@as(f64, @floatFromInt(discriminant))));
    while ((root + 1) <= discriminant / (root + 1)) : (root += 1) {}
    while (root > discriminant / root) : (root -= 1) {}
    return root * root == discriminant and (root + 1) % 6 == 0;
}

fn printNumber(value: anytype) void {
    var buffer: [64]u8 = undefined;
    const text = std.fmt.bufPrint(&buffer, "{d}\n", .{value}) catch return;
    _ = std.c.write(1, text.ptr, text.len);
}

File: classifier/data/raw/ai/samples/project_euler_java/Problem044.java
// Project Euler Problem 44: pentagonal numbers with pentagonal sum and difference.
public class Problem044 {
    public static void main(String[] args) {
        long best = Long.MAX_VALUE;
        // This bound contains the first (and minimal) qualifying pair.
        for (int k = 2; k <= 5_000; k++) {
            long pk = pentagonal(k);
            for (int j = k - 1; j >= 1; j--) {
                long pj = pentagonal(j);
                long difference = pk - pj;
                if (difference >= best) break;
                if (isPentagonal(difference) && isPentagonal(pk + pj)) best = difference;
            }
        }
        System.out.println(best);
    }

    private static long pentagonal(long n) {
        return n * (3 * n - 1) / 2;
    }

    private static boolean isPentagonal(long value) {
        long discriminant = 24 * value + 1;
        long root = (long) Math.sqrt(discriminant);
        return root * root == discriminant && (root + 1) % 6 == 0;
    }
}
